python/metatensor-learn/metatensor/learn/nn/module_map.py

```diff
@@ -0,0 +1,463 @@
+        ... )
+        >>> block_2 = TensorBlock(
+        ...     values=torch.tensor(
+        ...         [
+        ...             [5.0, 8.0, 2.0],
+        ...             [1.0, 2.0, 8.0],
+        ...         ]
+        ...     ),
+        ...     samples=Labels(
+        ...         ["structure", "center"],
+        ...         np.array(
+        ...             [
+        ...                 [0, 0],
+        ...                 [0, 1],
+        ...             ]
+        ...         ),
+        ...     ),
+        ...     components=[],
+        ...     properties=Labels(["properties"], np.array([[0], [1], [2]])),
+        ... )
+        >>> keys = Labels(names=["key"], values=np.array([[0], [1]]))
+        >>> tensor = TensorMap(keys, [block_1, block_2])
+        >>> linear = torch.nn.Linear(3, 1, bias=False)
+        >>> with torch.no_grad():
+        ...     _ = linear.weight.copy_(torch.tensor([1.0, 1.0, 1.0]))
+        ...
+        >>> # you could also extend the module by some nonlinear activation function
+        >>> from metatensor.learn.nn import ModuleMap
+        >>> module_map = ModuleMap.from_module(tensor.keys, linear)
+        >>> out = module_map(tensor)
+        >>> out[0].values
+        tensor([[ 7.],
+                [14.]], grad_fn=<MmBackward0>)
+        >>> out[1].values
+        tensor([[15.],
+                [11.]], grad_fn=<MmBackward0>)
+        """
+        module = deepcopy(module)
+        modules = []
+        for _ in range(len(in_keys)):
+            if many_to_one:
+                modules.append(module)
+            else:
+                modules.append(deepcopy(module))
+
+        return cls(in_keys, modules, out_properties)
+
+    def forward(self, tensor: TensorMap) -> TensorMap:
+        """
+        Takes a tensor map and applies the modules on each key it.
+
+        :param tensor:
+            input tensor map
+        """
+        out_blocks: List[TensorBlock] = []
+        for key, block in tensor.items():
+            out_block = self.forward_block(key, block)
+
+            for parameter, gradient in block.gradients():
+                if len(gradient.gradients_list()) != 0:
+                    raise NotImplementedError(
+                        "gradients of gradients are not supported"
+                    )
+                out_block.add_gradient(
+                    parameter=parameter,
+                    gradient=self.forward_block(key, gradient),
+                )
+            out_blocks.append(out_block)
+
+        return TensorMap(tensor.keys, out_blocks)
+
+    def forward_block(self, key: LabelsEntry, block: TensorBlock) -> TensorBlock:
+        module = self.get_module(key)
+
+        out_values = module.forward(block.values)
+        if self._out_properties is None:
+            properties = Labels.range("_", out_values.shape[-1])
+        else:
+            properties = self._out_properties[key]
```

Comment:
This will not work in general: if the `key` comes from a different Labels than the one used to initialize `self._out_properties`, this will raise KeyNotFound.

Reply:
Yes that was still something I was waiting to fix until comments from Arslan. Should work now!